source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -160,22 +180,52 @@ FilterDataStatus Filter::onData(ProcessorState& state, Buffer::Instance& data, b
       // The body has been buffered and we need to send the buffer
       ENVOY_LOG(debug, "Sending request body message");
       state.addBufferedData(data);
-      sendBodyChunk(state, *state.bufferedData(), true);
+      sendBodyChunk(state, *state.bufferedData(),
+                    ProcessorState::CallbackState::BufferedBodyCallback, true);
       // Since we just just moved the data into the buffer, return NoBuffer
       // so that we do not buffer this chunk twice.
+      state.setPaused(true);
       result = FilterDataStatus::StopIterationNoBuffer;
       break;
     }
 
     ENVOY_LOG(trace, "onData: Buffering");
+    state.setPaused(true);
     result = FilterDataStatus::StopIterationAndBuffer;
     break;
 
+  case ProcessingMode::STREAMED: {
+    switch (openStream()) {
+    case StreamOpenState::Error:
+      return FilterDataStatus::StopIterationNoBuffer;
+    case StreamOpenState::IgnoreError:
+      return FilterDataStatus::Continue;
+    case StreamOpenState::Ok:
+      // Fall through
+      break;
+    }
+
+    // Send the chunk on the gRPC stream
+    sendBodyChunk(state, data, ProcessorState::CallbackState::StreamedBodyCallback, end_stream);
+    // Move the data to the queue and optionally raise the watermark.
+    state.enqueueStreamingChunk(data, end_stream, true);
+
+    // At this point we will continue, but with no data, because that will come later
```

Comment:
Why do you continue with no data?

Reply:
We had some conversations about that in the issue that I opened before coding this up. I can find the original discussion, but without making big changes to the way that the filter chain is processed for all filters, there's no way for this filter to tell Envoy to stop sending new chunks of data right away, which in turn makes it hard to let an external processor modify each chunk individually. And even if we could do that, then it would introduce a lot of latency because we'd be pausing the filter flow after each chunk to wait for the processor.

This way, there is a flow of data to the external processor, and a flow of data back. We do this by moving the data out of each chunk and letting it keep flowing through the filter chain, and then introducing new chunks into the filter chain when they come back from the processor.

Reply:
This is complicated enough that I think linking to the thread in the comment would make sense; it's counter-intuitive that we would `Continue` (meaning pass to the next filter in the pipeline) without a chunk.

Reply:
Added a longer comment with a link to the issue.